### DISCUSSION TITLE
Support multi-domain configurations

### DISCUSSION THREAD

**Discussion by james-pre** [withastro/roadmap#702] [Proposal]
### Body

# TL;DR:

Astro could support multiple domains at runtime, with a single config file and build.

# Rationale

In many cases, a user may want to expose multiple sites with different domains using a single Astro project. For example, a user may want to create a collection of blogs at `blog.example.com`, documentation at `docs.example.com`, a main website at `example.com`, and expose an API at `api.example.com`. While it is possible to make the blogs and documentation available under the main website, it takes away flexibility from the user. Furthermore, the user can't expose an API since Astro does not support multiple domains.

I've been trying to do this myself and failed with these 2 examples, which could work with some minor changes to Astro.
Example #1:
```js
import express from 'express';
import { handler as ssrHandler } from '../dist/example.com/dist/server/entry.mjs';
import { handler as apiHandler } from '../dist/api.example.com/dist/server/entry.mjs';
import compression from 'compression';

const app = express();
app.use(compression());
app.use('/', express.static('dist/example.com/dist/client/'));
app.use(ssrHandler);
app.use('/', express.static(`dist/api.example.com/dist/client/`));
app.use(apiHandler);

app.listen(3000);
``` 

Example #2:
```js
import express from 'express';
import { handler as ssrHandler } from '../dist/example.com/dist/server/entry.mjs';
import { handler as apiHandler } from '../dist/api.example.com/dist/server/entry.mjs';
import compression from 'compression';

const handlers = new Map();
handlers.set('default', ssrHandler);
handlers.set('example.com', ssrHandler);
handlers.set('api.example.com', apiHandler);

const app = express();
app.use(compression());
app.all('*', (req, res) => {
	if(handlers.has(req.hostname)){
		const handler = handlers.get(hostname);
		handler(req, res);
	} else {
		res.status(404).send('Domain not found');
	}
});

app.listen(3000);
```

# Solution

Instead of using 2 different Astro configurations and builds, it would be very convenient to configure multiple domains in an Astro project. This could be easily done with a new config option, `sites`, which could be an array of `AstroUserConfig` or a subset of `AstroUserConfig`. Each configuration object could specify a different source directory, site, redirects, etc. Astro could server multiple domains at runtime by checking the request's host header against the configured domains. Astro could also support specifying a default behavior if the specified domain does not exist (e.g. the main site or an error page) and could support subdomains of localhost for development and such (perhaps using an `alias` config option, allowing the user to specify other possible base domains).

For example:
```js
import { defineConfig } from 'astro/config';

export default defineConfig({
	output: 'server',
	sites: [
		{
			site: 'example.com',
			srcDir: 'src/example.com',
			redirects: {
				'/docs': 'https://docs.example.com'
			},
			compressHTML: true
		},
		{
			site: 'api.example.com',
			srcDir: 'src/api.example.com',
		},
		{
			site: 'docs.example.com',
			srcDir: 'src/docs.example.com',
			alias: ['doc.example.com']
		},
	],
});
```

This is similar to an Apache VirtualHost.

**Comment by lilnasy** [comment]
Do you think https://github.com/withastro/roadmap/discussions/681 will cover this use case?

**Reply by james-pre** [reply]
I don't think #681 would fully cover this issue since it does support different config options (e.g. `compressHTML` from the example above). It looks awesome and would solve my personal use case though.

**Comment by codemaestro** [comment]
Hey all...  just wondering if there is any documented way to configure a single project to deploy different sites using shared components? I found some ideas generated by LLMs but can't find official documentation to show how to set up and configure a project for this use case.
